File: packages/qnsi/src/ai.ts
/**
 * QNSP AI Orchestrator - model registry, AI workloads with enclave
 * attestation, inference, bias / prompt-injection monitoring. Wraps
 * `apps/ai-orchestrator` (`/ai/v1`).
 */

import type { Internal, RequestOptions } from "./_internal.js";

const PATH_PREFIX = "/proxy/ai/v1";

export type ModelProvider =
	| "openai"
	| "anthropic"
	| "google"
	| "meta"
	| "deepseek"
	| "bedrock"
	| "huggingface"
	| "custom";

export type ModelType =
	| "llm"
	| "embedding"
	| "image"
	| "audio"
	| "multimodal"
	| "classification"
	| "regression";

export type WorkloadPriority = "low" | "normal" | "high";
export type SchedulingPolicy = "spot" | "on-demand" | "mixed";

export interface RegisterModelRequest {
	readonly name: string;
	readonly version: string;
	readonly provider: ModelProvider;
	readonly modelType: ModelType;
	readonly description?: string;
	readonly baseModelId?: string;
	readonly capabilities?: readonly string[];
	readonly inputSchema?: Record<string, unknown>;
	readonly outputSchema?: Record<string, unknown>;
	readonly trainingMetadata?: {
		readonly datasetId?: string;
		readonly epochs?: number;
		readonly learningRate?: number;
		readonly trainedAt?: string;
	};
	readonly servingConfig?: {
		readonly maxBatchSize?: number;
		readonly maxConcurrency?: number;
		readonly timeoutMs?: number;
		readonly cachingEnabled?: boolean;
		readonly streamingEnabled?: boolean;
	};
	readonly tags?: readonly string[];
}

export interface UpdateModelRequest {
	readonly description?: string;
	readonly capabilities?: readonly string[];
	readonly inputSchema?: Record<string, unknown>;
	readonly outputSchema?: Record<string, unknown>;
	readonly servingConfig?: RegisterModelRequest["servingConfig"];
	readonly performanceMetrics?: {
		readonly avgLatencyMs?: number;
		readonly p99LatencyMs?: number;
		readonly throughputRps?: number;
		readonly errorRate?: number;
	};
	readonly tags?: readonly string[];
}

export interface DeployModelRequest {
	readonly modelId: string;
	readonly environment: "development" | "staging" | "production";
	readonly replicas?: number;
	readonly minReplicas?: number;
	readonly maxReplicas?: number;
	readonly resourceAllocation?: {
		readonly cpu?: number;
		readonly memoryGiB?: number;
		readonly gpu?: number;
		readonly acceleratorType?: string;
	};
	readonly autoscalingConfig?: {
		readonly targetCpuUtilization?: number;
		readonly targetMemoryUtilization?: number;
		readonly scaleUpCooldownSeconds?: number;
		readonly scaleDownCooldownSeconds?: number;
	};
	readonly healthCheckConfig?: {
		readonly path?: string;
		readonly intervalSeconds?: number;
		readonly timeoutSeconds?: number;
		readonly unhealthyThreshold?: number;
	};
}

export interface WorkloadResources {
	readonly cpu: number;
	readonly memoryGiB: number;
	readonly gpu: number;
	readonly acceleratorType: string;
}

export interface SubmitWorkloadRequest {
	readonly name: string;
	readonly priority: WorkloadPriority;
	readonly schedulingPolicy: SchedulingPolicy;
	readonly containerImage: string;
	readonly command: readonly string[];
	readonly env: Record<string, string>;
	readonly resources: WorkloadResources;
	readonly artifacts: readonly {
		readonly artifactId: string;
		readonly mountPath: string;
		readonly accessMode: "read" | "read-write";
	}[];
	readonly manifest: {
		readonly algorithm: string;
		readonly pqcSignature: string;
		readonly issuedAt: string;
		readonly [key: string]: unknown;
	};
	readonly labels?: Record<string, string>;
}

export interface InferenceRequest {
	readonly modelDeploymentId: string;
	readonly input: Record<string, unknown>;
	readonly provider?: string;
	readonly priority?: WorkloadPriority;
	readonly schedulingPolicy?: SchedulingPolicy;
	readonly command?: readonly string[];
	readonly env?: Record<string, string>;
	readonly resources?: Partial<WorkloadResources>;
}

export interface RegisterArtifactRequest {
	readonly documentId: string;
	readonly version: number;
}

export class AiClient {
	constructor(private readonly internal: Internal) {}

	// ── Model registry ──────────────────────────────────────────────

	registerModel(req: RegisterModelRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/registry/models`, req, opts);
	}

	listModels(query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/registry/models`, undefined, { query });
	}

	getModel(modelId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/registry/models/${modelId}`);
	}

	updateModel(
		modelId: string,
		body: UpdateModelRequest,
		opts?: Pick<RequestOptions, "idempotencyKey">,
	) {
		return this.internal.request("PATCH", `${PATH_PREFIX}/registry/models/${modelId}`, body, opts);
	}

	activateModel(modelId: string, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request(
			"POST",
			`${PATH_PREFIX}/registry/models/${modelId}/activate`,
			{},
			opts,
		);
	}

	deployModel(body: DeployModelRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/registry/deployments`, body, opts);
	}

	// ── Workloads ───────────────────────────────────────────────────

	submitWorkload(req: SubmitWorkloadRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/workloads`, req, opts);
	}

	getWorkload(workloadId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/workloads/${workloadId}`);
	}

	listWorkloads(query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/workloads`, undefined, { query });
	}

	cancelWorkload(workloadId: string, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request(
			"POST",
			`${PATH_PREFIX}/workloads/${workloadId}/cancel`,
			undefined,
			opts,
		);
	}

	// ── Inference ────────────────────────────────────────────────────

	invokeInference(req: InferenceRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/inference`, req, opts);
	}

	// ── Artifacts ────────────────────────────────────────────────────

	registerArtifact(req: RegisterArtifactRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/artifacts`, req, opts);
	}
}

File: sdks/rust/qnsi/src/ai.rs
//! QNSI AI Orchestrator - model registry, AI workload submission with
//! enclave attestation, inference, and bias / prompt-injection monitoring.

use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::Arc;

use crate::activation::Activation;
use crate::errors::Error;
use crate::service_client::ServiceClient;

const PATH_PREFIX: &str = "/proxy/ai/v1";

#[derive(Clone)]
pub struct Client {
    sc: ServiceClient,
}

impl Client {
    pub fn new(activation: Arc<Activation>, http: reqwest::Client) -> Self {
        Self { sc: ServiceClient::new(activation, http, PATH_PREFIX) }
    }

    // ── Model registry ──────────────────────────────────────────────

    pub async fn register_model(&self, req: RegisterModelRequest, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::POST, "/models", Some(&req), None, idempotency_key).await
    }

    pub async fn list_models(&self, query: &[(&str, String)]) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, "/models", None, Some(query), None).await
    }

    pub async fn get_model(&self, model_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/models/{}", model_id), None, None, None).await
    }

    pub async fn update_model(&self, model_id: &str, body: serde_json::Map<String, Value>, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc
            .request(Method::PATCH, &format!("/models/{}", model_id), Some(&Value::Object(body)), None, idempotency_key)
            .await
    }

    pub async fn activate_model(&self, model_id: &str, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc
            .request::<()>(Method::POST, &format!("/models/{}/activate", model_id), None, None, idempotency_key)
            .await
    }

    pub async fn deploy_model(&self, body: serde_json::Map<String, Value>, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc
            .request(Method::POST, "/models/deploy", Some(&Value::Object(body)), None, idempotency_key)
            .await
    }

    // ── Workloads ───────────────────────────────────────────────────

    pub async fn submit_workload(&self, req: SubmitWorkloadRequest, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::POST, "/workloads", Some(&req), None, idempotency_key).await
    }

    pub async fn get_workload(&self, workload_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/workloads/{}", workload_id), None, None, None).await
    }

    pub async fn list_workloads(&self, query: &[(&str, String)]) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, "/workloads", None, Some(query), None).await
    }

    pub async fn cancel_workload(&self, workload_id: &str, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc
            .request::<()>(Method::POST, &format!("/workloads/{}/cancel", workload_id), None, None, idempotency_key)
            .await
    }

    // ── Inference ────────────────────────────────────────────────────

    pub async fn invoke_inference(&self, req: InferenceRequest, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::POST, "/inference", Some(&req), None, idempotency_key).await
    }

    // ── Artifacts ────────────────────────────────────────────────────

    pub async fn register_artifact(&self, req: RegisterArtifactRequest, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::POST, "/artifacts", Some(&req), None, idempotency_key).await
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisterModelRequest {
    pub name: String,
    pub version: String,
    pub provider: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Map<String, Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmitWorkloadRequest {
    #[serde(rename = "modelId")]
    pub model_id: String,
    #[serde(rename = "type")]
    pub workload_type: String,
    #[serde(rename = "inputRefs", skip_serializing_if = "Option::is_none")]
    pub input_refs: Option<Vec<String>>,
    #[serde(rename = "outputBucket", skip_serializing_if = "Option::is_none")]
    pub output_bucket: Option<String>,
    #[serde(rename = "enclaveType", skip_serializing_if = "Option::is_none")]
    pub enclave_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Map<String, Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InferenceRequest {
    #[serde(rename = "modelId")]
    pub model_id: String,
    pub input: serde_json::Map<String, Value>,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub stream: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Map<String, Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisterArtifactRequest {
    pub name: String,
    pub hash: String,
    #[serde(rename = "storageId")]
    pub storage_id: String,
    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
    pub artifact_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Map<String, Value>>,
}
